lib/cache/access_list_test.go

```diff
@@ -175,3 +178,76 @@ func TestAccessListReviews(t *testing.T) {
 		deleteAll: p.accessLists.DeleteAllAccessListReviews,
 	})
 }
+
+func TestCountAccessListMembersScoping(t *testing.T) {
+	modules.SetTestModules(t, &modules.TestModules{
+		TestBuildType: modules.BuildEnterprise,
+		TestFeatures: modules.Features{
+			Entitlements: map[entitlements.EntitlementKind]modules.EntitlementInfo{
+				entitlements.Identity: {Enabled: true},
+			},
+		},
+	})
+
+	p := newTestPack(t, ForAuth)
+	t.Cleanup(p.Close)
+
+	ctx := context.Background()
+	clock := clockwork.NewFakeClock()
+
+	alA, err := p.accessLists.UpsertAccessList(ctx, newAccessList(t, "list-a", clock))
+	require.NoError(t, err)
+	alB, err := p.accessLists.UpsertAccessList(ctx, newAccessList(t, "list-b", clock))
+	require.NoError(t, err)
+
+	var (
+		alaUsers = 4
+		albUsers = 3
+		alaLists = 2
+		albLists = 1
+	)
+
+	// add users to list a
+	for i := 0; i < alaUsers; i++ {
+		member := newAccessListMember(t, alA.GetName(), fmt.Sprintf("user-%d", i))
+		member.Spec.MembershipKind = accesslist.MembershipKindUser
+		_, err := p.accessLists.UpsertAccessListMember(ctx, member)
+		require.NoError(t, err)
+	}
+	for i := 0; i < albUsers; i++ {
+		member := newAccessListMember(t, alB.GetName(), fmt.Sprintf("b-user-%d", i))
+		member.Spec.MembershipKind = accesslist.MembershipKindUser
+		_, err := p.accessLists.UpsertAccessListMember(ctx, member)
+		require.NoError(t, err)
+	}
+
+	for i := 0; i < alaLists; i++ {
+		nestedList, err := p.accessLists.UpsertAccessList(ctx, newAccessList(t, fmt.Sprintf("nested-list-a-%d", i), clock))
+		require.NoError(t, err)
+		member := newAccessListMember(t, alA.GetName(), nestedList.GetName())
+		member.Spec.MembershipKind = accesslist.MembershipKindList
+		_, err = p.accessLists.UpsertAccessListMember(ctx, member)
+		require.NoError(t, err)
+	}
+	for i := 0; i < albLists; i++ {
+		nestedList, err := p.accessLists.UpsertAccessList(ctx, newAccessList(t, fmt.Sprintf("nested-list-b-%d", i), clock))
+		require.NoError(t, err)
+		member := newAccessListMember(t, alB.GetName(), nestedList.GetName())
+		member.Spec.MembershipKind = accesslist.MembershipKindList
+		_, err = p.accessLists.UpsertAccessListMember(ctx, member)
+		require.NoError(t, err)
+	}
+
+	// wait for cache to reflect updates
```

Comment:
I don't know how far you want to push this but waiting for 150msec is sort of a lot; the test pack object has an `eventsC` field with a channel that gets fed by the cache whenever events happen, you could use that to rerun the check whenever something changes in the cache rather than just blocking for time.

Reply:
To clarify, I was thinking of something like the following

```go
	// wait for cache to reflect updates
	timeout := time.After(5 * time.Second)
	for i := 0; ; i++ {
		if i > 0 {
			select {
			case <-timeout:
				require.Fail(t, "timeout")
			case <-p.eventsC:
			}
		}
		check here, require.NoError on CountAccessListMembers and continue if the numbers are wrong
		break
	}
```

Reply:
Gotcha, thanks! Just copied the loop from the existing test in this file; this looks like a much better solution.